### PR TITLE
[FLINK-9852] [table] Expose descriptor-based sink creation

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This commit exposes the new unified sink creation through the table environments and the external catalog table. It introduce a new `update-mode` property in order to distinguish between append, retract, and upsert table sources and sinks. This commit refactors the top-level API classes a last time and adds more documentation. This commit completes the unified table sources/sinks story from an API point of view.

## Brief change log

- Introduction of `TableEnvironment.connect()` and corresponding API builder classes
- Introduction of property `update-mode:` and update of existing connectors
- External catalog support with proper source/sink discovery and API

## Verifying this change

Existing tests were adapted.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+      case _ =>
+        throw new ValidationException(
+          "External catalog table does not support the current environment for a table sink.")
+    }
+
+    new TableSourceSinkTable[T1, T2](source, sink)
+  }
+
+  private def createBatchTableSource[T](
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala]: I still do not like this lack of abstraction between batch and streaming in form of `createBatchTableSource`/`createStreamTableSource`.

Instead of writing if/elses everywhere in our code there should be some common layer that handles such logic. Here half of the problem boils down to factories with methods createBatchTableSource and createStreamTableSource.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/SchematicDescriptor.scala [hunk 2]
```diff
@@ -19,14 +19,17 @@
 package org.apache.flink.table.descriptors
 
 /**
-  * Common class for all descriptors describing a table sink.
+  * A trait for descriptors that allow to define a format and schema.
   */
-abstract class TableSinkDescriptor extends TableDescriptor {
+trait SchematicDescriptor extends Descriptor {
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/SchematicDescriptor.scala]: are you using it anywhere as interface? what does extracting it to separate interface give us? Maybe drop it?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    if (connectorDescriptor.needsFormat() && formatDescriptor.isEmpty) {
+      throw new ValidationException(
+        s"The connector '$connectorDescriptor' requires a format description.")
+    } else if (!connectorDescriptor.needsFormat() && formatDescriptor.isDefined) {
+      throw new ValidationException(
+        s"The connector '$connectorDescriptor' does not require a format description " +
+          s"but '${formatDescriptor.get}' found.")
+    }
+
+    // basic validation
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala]: drop the comment

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    this
+  }
+
+  override def toString: String = {
+    getValidProperties.toString
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  private def getValidProperties: DescriptorProperties = {
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala]: duplicated code with `StreamTableDescriptor. getValidProperties`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+package org.apache.flink.table.descriptors
+
+import org.apache.flink.table.api.{StreamTableEnvironment, ValidationException}
+import org.apache.flink.table.descriptors.StreamTableDescriptorValidator._
+import org.apache.flink.table.factories.{StreamTableSinkFactory, StreamTableSourceFactory, TableFactoryService}
+
+/**
+  * Descriptor for specifying a table source and/or sink in a streaming environment.
+  */
+class StreamTableDescriptor(
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala]: This class duplicates code with `BatchTableDescriptor`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  with SchematicDescriptor
+  with RegistrableDescriptor {
+
+  private var formatDescriptor: Option[FormatDescriptor] = None
+  private var schemaDescriptor: Option[Schema] = None
+
+  /**
+    * Internal method for properties conversion.
+    */
+  override private[flink] def addProperties(properties: DescriptorProperties): Unit = {
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala]: in many places (for example all `addProperties` methods) you are ordering methods very weirdly. Rule of thumb should be pubic methods before private. Longer story: https://stackoverflow.com/a/1760877/8149051

Many times in this code review I had to jump up & down.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/CsvAppendTableSinkFactory.scala [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.apache.flink.table.descriptors.SchemaValidator.{SCHEMA, SCHEMA_PROCTIME}
+import org.apache.flink.table.descriptors.StreamTableDescriptorValidator.{UPDATE_MODE, UPDATE_MODE_VALUE_APPEND}
+import org.apache.flink.table.factories.StreamTableSinkFactory
+import org.apache.flink.types.Row
+
+/**
+  * Factory base for creating configured instances of [[CsvTableSink]] in a stream environment.
+  */
+class CsvAppendTableSinkFactory
```
Comment (pnowojski) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/CsvAppendTableSinkFactory.scala]: This should be done in separate commit

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala [hunk 8]
```diff
@@ -18,33 +18,299 @@
...
+  *       .version("0.11"))
+  *   .withFormat(
+  *     new Json()
+  *       .jsonSchema("{...}")
+  *       .failOnMissingField(false))
+  *   .withSchema(
+  *     new Schema()
+  *       .field("user-name", "VARCHAR").from("u_name")
+  *       .field("count", "DECIMAL")
+  *   .asTableSource()
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala]: I also thought about that but actually descriptors don't "build" something. The only final representation would be the properties but we don't expose them to the user.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      externalCatalogTable: ExternalCatalogTable,
+      javaMap: util.Map[String, String],
+      statistics: FlinkStatistic)
+    : Option[TableSourceTable[T]] = if (externalCatalogTable.isTableSource) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala]: I also like this pattern more but it is not very Scala-like:
https://tpolecat.github.io/2014/05/09/return.html

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala (hunk 1 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala]: Then we would have 4 factories that have to be checked with if/else branches. Having those if/else at 3 places (SQL Client, external catalog, and descriptors) is acceptable in my opinion.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/SchematicDescriptor.scala (hunk 2 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/SchematicDescriptor.scala]: `SchematicDescriptor` is used for `ExternalCatalogTable`, `StreamTableDescriptor`, and `BatchTableDescriptor`. If we add a new level next to `connector`, `format` (which may happens in the future), we would immediately get a compile error there.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala (hunk 3 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala]: I extended the comment.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala (hunk 4 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala]: I agree but I cannot move it up the class hierarchy because it would be public in Java. I will create a util class.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala (hunk 6 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala]: Depends on the language you are using. This method is public in Java ;)
But I will move it down.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala (hunk 5 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala]: I don't see an alternative if we don't want to have an ugly API.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/CsvAppendTableSinkFactory.scala (hunk 7 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/CsvAppendTableSinkFactory.scala]: Will put it into a separate commit while merging.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala (hunk 5 above)
Comment (pnowojski) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/StreamTableDescriptor.scala]: Inherit:
```
class StreamTableDescriptor(...) extends BatchTableDescriptor with StreamableDescriptor
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala [hunk 10]
```diff
@@ -18,33 +18,299 @@
...
-  this.connectorDescriptor = Some(connectorDesc)
-  this.formatDescriptor = formatDesc
-  this.schemaDescriptor = schemaDesc
-  this.statisticsDescriptor = statisticsDesc
-  this.metaDescriptor = metadataDesc
-
-  // expose statistics for external table source util
-  override def getTableStats: Option[TableStats] = super.getTableStats
+class ExternalCatalogTable(val connectorDescriptor: ConnectorDescriptor)
+  extends TableDescriptor
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala]: I would say yes. See also `org.apache.flink.runtime.jobmanager.JobManager` as an example.

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala]: This is the code style that we should all comply with.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala]: A descriptor is basically a builder. That builds immutable string properties ones the `addProperties` is called.

"format/schema/statistics/metadata/updateMode" are only concepts that we propose to a user. In the end, a user can implement a custom connector descriptor and paste whatever properties he would like to have and implement a custom factory that match to whatever properties are needed.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableUtil.scala]: I will reverse the branches.
